assertj-core/src/main/java/org/assertj/core/internal/AbstractComparisonStrategy.java

```diff
@@ -39,7 +40,20 @@ public Iterable<?> duplicatesFrom(Iterable<?> iterable) {
         noDuplicates.add(element);
       }
     }
-    return duplicates;
+    return duplicatesWithOrderPreserved(iterable, duplicates);
+  }
+
+  private Iterable<?> duplicatesWithOrderPreserved(Iterable<?> iterable, Set<Object> duplicates) {
+    Set<Object> duplicatesInOriginalOrder  = new LinkedHashSet<>();
+    // for elements of type Set collections, ensure each element is added only once
+    Set<Object> alreadyAddedDuplicates = newSetUsingComparisonStrategy();
+    for (Object element : iterable) {
+      if (duplicates.contains(element)&& !alreadyAddedDuplicates.contains(element)) {
```

Comment:
you can get rid of addedAddDuplicates entirely, if you make the second part of the if test `!duplicatesInOriginalOrder.contains(element)`

Reply:
For elements of array type e.g. newArrayList(array("Merry"), array("Frodo"),  array("Merry")) TreeSet uses custom comparator to identify the duplicates. This is been done by checking the deeply equals for two array types( areEquals method). So elements array("Merry") and array("Merry") will be treated as duplicates.

since we don't have this comparator in LinkedHashSet, it will not be able to identify two arrays with same elements as duplicates.

Reply:
I might be missing a lot of context here, so ignore me if that is the case. Instead of introducing a private method, can this be done in the `duplicatesFrom` method instead, like replacing Lines 34 with a LinkedHashSet?

Reply:
earlier it was not possible but with switch option-2 (first duplicate first in response) implementation its possible. 

I have made that change.